Initial commit for PyTorch to run on AMD hardware using the ROCM software stack

wsttiger opened the pull request:
This is the initial commit which adds a new AMD hardware target .... more to come. @soumith

wsttiger committed:
Initial commit for ROCM stack configuration files

wsttiger committed:
Initial commit for ROCM stack configuration files

wsttiger committed:
Disabled sparse functions on the ROCM path

wsttiger committed:
Disabled sparse functions on the ROCM path

wsttiger committed:
Translated a few kernels to HIP-compatible forms on the ROCM stack

wsttiger committed:
Translated a few kernels to HIP-compatible forms on the ROCM stack

wsttiger committed:
Removed make_magic_wrapper call in THC

wsttiger committed:
Removed make_magic_wrapper call in THC

ezyang committed:
Top level comment for gen.py

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Delete obsolete only_registry entries in cwrap.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Support defining gradient for multiple inputs simultaneously in derivatives.yaml

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

ezyang committed:
Make atan2 backwards reuse intermediate computation.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

chenyuntc committed:
ignore digit in container's __dir__

ezyang committed:
Some documentation for derivatives.yaml

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

wsttiger committed:
Fixed spaces in setup.py

wsttiger committed:
Fixed spaces in setup.py

soumith commented on aten/src/THCS/generic/THCSTensor.c:
```diff
@@ -176,62 +205,86 @@ THCSTensor *THCSTensor_(newWithTensorAndSize)(THCState *state, THCIndexTensor *i
 
   return self;
+#else
+  return NULL;
```
can you change the #else conditions to `THError("not supported on HIP_PLATFORM")`

wsttiger replied:
Sure

soumith commented on aten/src/THC/generic/THCTensorIndex.cu:
```diff
@@ -240,18 +258,35 @@ void THCTensor_(indexAdd)(THCState *state, THCTensor *dst, int dim, THCudaLongTe
+        dstAddDim, srcAddDim, sliceSize, dstAddDimSize);
+  
+  #define LARGE_INDEX(TENSOR_TYPE, TYPE, DST_DIM, SRC_DIM, IDX_DIM) \
+    hipLaunchKernelGGL(                                             \
```
why is the hibLaunchKernelGGL needed here? can you give some context, so that we can see if we can work out a solution that doesn't need such a special-case

wsttiger replied:
Unfortunately our parsing tool has issues with converting kernel launches inside multiline macros. Therefore, for these cases, it was necessary to include a separate compilation using the HIP-specific preprocessor guards.

colesbury committed:
Fix float uniform generation in TH (#3541)

Generate random uniform floats in the range [0, 1) by generating random
uniform uint32 in the range [0, 2^24-1] and dividing by 2^24. This
ensures that the largest value is representable as a float32 less than
one.

This also changes the uniform double generation to use more bits of
randomness.

Commit:
avoid unnecessary multiplies in derivatives (#3545)

ssnl committed:
Exposing emptyCache from allocator (#3518)

* Add empty_cache binding

* cuda.empty_cache document

* update docs

gchanan committed:
Generate native functions with const ref Tensor arguments. (#3465)

* Generate native functions with const ref Tensor arguments.

This matches the non-native functions and avoids unnecessary ref counts.

* Properly handle inplace functions.

* Return Tensor & for inplace native functions.

zou3519 committed:
Previous PyTorch version info (#3549)

wsttiger committed:
Added error message for unsupported functionality

wsttiger committed:
Added error message for unsupported functionality

gchanan committed:
Have ATen build respect DEBUG variable.

gchanan committed:
Prevent segfaults from undefined aten tensors (#3482)

* Prevent segfaults from undefined aten tensors.

This introduces a singleton UndefinedTensor TensorImpl with UndefinedType that is the starting state of a Tensor with no constructor arguments.  In this way we avoid null pImpls and avoid segfaults
without having to if-check each pImpl dereference.

* If either Backend or Scalar type is Undefined in registry, return
the UndefinedType to avoid errors like CPUUndefinedType is not enabled.

* Address review comments.

* Avoid refcounting UndefinedTensors.

* Use reference_wrapper to avoid copy in check_defined.

* Declare UndefinedTensor singleton as class-static.

* Seperate checked_cast into storage and tensor versions.

* Include <functional>

* Handle nullptr TensorImpls coming from NN.

* Fix nullptr check in batch_normalization backward with defined check.

wsttiger committed:
fixed setup.py for pylint

wsttiger committed:
fixed setup.py for pylint

Commit:
add numpy() and from_numpy() to HalfTensor (#2953)

colesbury committed:
Fix stride checks in gemm dispatch (#3548)

From https://software.intel.com/en-us/mkl-developer-reference-fortran-gemm:

 lda: "When transa = 'N' or 'n', then lda must be at least max(1, m),
       otherwise lda must be at least max(1, k)."

 ldb: "When transb = 'N' or 'n', then ldb must be at least max(1, k),
       otherwise ldb must be at least max(1, n)."

Partly addresses #3525

colesbury committed:
Fix uniform on CUDA tensor to return in range [0, 1) (#3547)

The curand_uniform function returns the range (0, 1]. Most RNG APIs have
the opposite bounds. Fixup the values in uniform_() so that they fall in
the more common bounds.

zou3519 committed:
Fix error when default_collate is passed a collection of numpy.str_ (#3404)

* Fix error when default_collate is passed a collection of numpy.str_

* Error if default_collate input is nested nparray containing non-numbers

gchanan committed:
[ATen][Scalars] Remove Scalar from return types of functions. (#3557)

* Add direct C-type scalar conversions from Tensor, e.g. toCFloat() as an alias for Scalar(x).toFloat()

* Provide tensor overloads for fill_, masked_fill_, index_fill_.

* Everythign up to scalar overload.

* Fix pytorch build for aten scalar return type changes.

* Use valid expression instead of dangling else.

* Simplify code generation.

* Fix test_jit (why didn't this compile locally?)

peterjc123 committed:
Improve Windows Compatibility (for csrc/scripts) (#2941)

wsttiger committed:
more setup.py changes
